src/Npgsql/PoolManager.cs

```diff
@@ -91,11 +92,14 @@ internal static void Clear(string connString)
 
         internal static void ClearAll()
         {
-            for (var i = 0; i < _nextSlot; i++)
+            lock (_lock)
             {
-                if (_pools[i].Key == null)
-                    return;
-                _pools[i].Pool?.Clear();
+                for (var i = 0; i < _nextSlot; i++)
+                {
+                    if (_pools[i].Key == null)
+                        return;
+                    _pools[i].Pool?.Clear();
+                }
```

Comment:
I think that a local variable could be introduced before the loop to store `_pools`. Otherwise, there will be a lot of memory barriers.

Reply:
Okay, I've added your suggestions in the latest commit. Thanks!

Reply:
You could also use `var cp = pools[i];` inside the for loop to minimize array lookups:
```
var cp = pools[i];
if (cp.Key == null)
    return;
cp.Pool?.Clear();
```
Would look at the other places in `PoolManager.cs` for this optimization as well.